text/0000-cargo-rustup-discovery.md

```diff
@@ -0,0 +1,482 @@
+
+#### Current git discovery
+
+The following is a list of places where Cargo uses git repositories:
+
+- Uses repo path discovery:
+    - Package listing: Determines the list of files in the package.
+        - Used in several places:
+            - `cargo package`, `cargo publish`
+            - `cargo vendor`
+            - Checking for the "newest" file for fingerprinting (`cargo doc` and build scripts).
+        - Cargo does not use git if `package.include` is set.
+        - Currently Cargo ignores errors opening the repository, and falls back to a dumb walk.
+            - This may be a hazard because it could silently change which files are included during publishing. Disables dirty checks and `.cargo_vcs_info.json`.
+    - `cargo fix`: Checks for dirty files.
+        - If Cargo fails to open the repo, it will fail requiring `--allow-no-vcs`.
+    - `cargo new`:
+        - Uses discovery to determine if already inside a git repository. Failure is silently ignored and assumes that it is not in a pre-existing repository.
+        - Uses libgit2 to initialize a new repository.
+- Loads directly from a specific path:
+    - Loading the registry index.
+        - Currently Cargo will silently try to delete the index if it fails to open, and reinitialize it. If there is a permission error, it will fail.
+    - Loading a git dependency.
+        - If Cargo fails to open the repo, then it will try to delete it and reinitialize and fetch it.
+    - Fetching behavior:
+        - Cargo may run `git gc` based on a heuristic. If `git gc` fails, Cargo will try to delete the repo and reinitialize it.
+          This may present a hazard if a user is using git 2.35.2 or newer, where it may fail with the ownership check (such as running as root with an inherited `HOME`).
+          The consequence is that cargo will have to refetch the git repo when it wouldn't normally.
+
+# Appendix
+
+## Git behavior
+
+In response to [CVE-2022-24765], git introduced ownership checks and a
+[`safe.directory`] configuration option to override them.
+More about this change is described in the [GitHub blog post].
+
+`git` searches from the current directory up to the filesystem root to find a git repo.
+This is a vulnerability because the repo may contain config settings that execute arbitrary programs (particularly [`core.fsmonitor`]).
+If an attacker has the ability to write to directories at or above your directory, then they can use it to escalate.
+
+A classic example is `C:\` on Windows which in some environments is writeable by all users.
+
+This is particularly problematic as many users place `git` commands in their shell's prompt,
+causing it to open repositories from every directory the user enters.
+
+`git` has been changed so that it will not open a repository owned by a different user, and instead return an error.
+For most cases, it will check that the ownership of the directory *containing* the `.git` directory matches the current user.
```

Comment:
Might worth mentioning that a fix for [CVE-2022-29187](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2022-29187) (missing part of [CVE-2022-24765](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2022-24765)) was released two weeks ago: https://github.com/git/git/commit/3b0bf2704980b1ed6018622bdf5377ec22289688

Basically, git now takes into account the ownership of gitdir and gitfile as well. Unlike git, this RFC proposes to check files themselves, so I don't think we need a revise accordingly.